### ISSUE TITLE
_zsh_autosuggest_strategy_default:5: error in flags

### ISSUE THREAD

**Issue by sysrich** [zsh-users/zsh-autosuggestions#296]
I am running zsh-autosuggestions from git/master

Since this morning all of my zsh shells are unusable because of the error message in $sub

Reverting to zsh-autosuggestions from earlier than 0.4.0 seems to fix it

Any suggestions on how to fix it so I can run zsh-autosuggestions latest again?

**Comment by nomike** [comment]
Also affects me

**Comment by ericfreese** [comment]
What is the error message? What version of zsh? What OS?

**Comment by ericfreese** [comment]
Also, do you get errors running these?

```shell
$ echo ${history[(r)ls*]}
$ foo="bar"; echo ${(b)foo}
```

**Comment by chenxinle** [comment]
centos7, Manual install or Oh My Zsh install both has the problem.
zsh version is 0.4.1.
help!

**Comment by SunMar** [comment]
Same issue here with 0.4.1, running it as an oh-my-zsh plugin. Reverting to 0.4.0 solved the issues.

Running your second line gives an error:

```
$ echo ${history[(r)ls*]}
ls -l
$ foo="bar"; echo ${(b)foo}
zsh: error in flags
```

Amazon Linux 2017.09, latest OhMyZsh (commit e273cf004e1ff0510aee61416885f3003bcd15d9), zsh 5.0.2

**Comment by sysrich** [comment]
Distribution - openSUSE Leap 42.3 and openSUSE Tumbleweed

zsh version 5.0.2 (Leap) or 5.3.1 (Tumbleweed)

Reproduced with zsh -f and sourcing `zsh-autosuggestions` manually

````
luke% echo ${history[(r)ls*]}
luke% foo="bar"; echo ${(b)foo}
_zsh_autosuggest_strategy_default:5: error in flags
luke% foo="bar"; echo ${(b)foo}
````

zshrc https://github.com/sysrich/salt-states/blob/master/opensuse/zshrc-user

Please help, two days of no zsh is making me a sad bash-using panda 🐼

**Comment by ericfreese** [comment]
Ok, thanks for the additional info. It looks like support for the `(b)` flag was added relatively recently (released with zsh v5.0.8):

https://github.com/zsh-users/zsh/commit/6269db883ad6989a578bac299deab9c4285bb1df

This plugin doesn't currently support zsh versions older than v5.0.8: https://github.com/zsh-users/zsh-autosuggestions/blob/master/circle.yml#L3

The `(b)` flag was added to this plugin in commits 7f8ff28 and 5e4487e as a simplification of the pattern character escaping logic used when looking up suggestions. It also fixed some issues with the `match_prev_cmd` suggestion strategy.

The easiest fix here seems to be for you to upgrade zsh to v5.0.8 or higher, but if for some reason you can't, using v0.4.0 is always available as a workaround. You could also define your own strategy function that doesn't use the `(b)` flag.

That being said, if a lot of people are unhappy with the workarounds and for some reason are not able to upgrade to v5.0.8 or higher, we could consider adding back the code that was removed by using the `(b)` flag and officially supporting versions of zsh older than v5.0.8. Toward that end, if anyone is stuck on a version older than 5.0.8 and isn't able to upgrade, can you add a note here with the version you're stuck on and an explanation of why you can't upgrade?

**Comment by sysrich** [comment]
I'm running Tumbleweed, with zsh 5.3.1, and yet still having this problem. How can your suggestion that the solution to this is upgrading to 5.0.8?

**Comment by ericfreese** [comment]
Can you double check that you're on 5.3.1? I can't imagine how this would be an issue with that version of zsh. From your shell session, you can `echo $ZSH_VERSION`.

**Comment by sysrich** [comment]
No I can't, because `echo $ZSH_VERSION` results in `_zsh_autosuggest_strategy_default:5: error in flags` ;)

`zypper if zsh` however results in:

````
zypper if zsh
Loading repository data...
Reading installed packages...


Information for package zsh:
----------------------------
Repository     : openSUSE-20150318-0                
Name           : zsh                                
Version        : 5.3.1-1.6
````

**Comment by ericfreese** [comment]
> No I can't, because echo $ZSH_VERSION results in _zsh_autosuggest_strategy_default:5: error in flags ;)

To get around this, you can temporarily remove zsh-autosuggestions from your zshrc or run `zsh -f`. Looking into the details of the openSUSE zsh package now.

**Comment by mphj** [comment]
I've same problem in ubuntu 14.04 zsh 5.0.2

**Comment by SunMar** [comment]
As noted before I am running on 5.0.2 and can't upgrade because that's the version provided by our distro (Amazon Linux). Manually upgrading is not an option due to compliance with security policies. In a business setting it will probably be very common to still run on 5.0.2 and it will not be possible to upgrade. :-(

**Comment by nomike** [comment]
I'm running Redhat Enterprise Linux 6 which has zsh version 4.3.11.
I can't switch to any newer version as this is not covered by support and there are no official packages available for RHEL6 and will never be.
On RHEL7 the zsh version is 5.0.2 and might at some day get an update to 5.0.8, but Redhat is quite conservative on this (for good reasons).

**Comment by sysrich** [comment]
Problem also reproduced on SUSE Linux Enterprise Server and Desktop 12, another distro where updating to 5.0.8 is unlikely for a few years

**Comment by ikirmitz** [comment]
Same for me on Debian Wheezy with zsh 5.4.2 compiled from source.

**Comment by ericfreese** [comment]
@ikirmitz Interesting. Does your compiled zsh support the `(b)` flag?

What is the output of these commands?

```
$ echo $ZSH_VERSION
$ foo="bar"; echo ${(b)foo}
```

**Comment by ikirmitz** [comment]
Hi @ericfreese . Upon further checks, it turns out that although which zsh pointed to the one I compiled from source, $SHELL was pointing in an older version installed in /etc/alternatives (4.3.17).
I fixed that and now have no issues. Sorry for the confusion

**Comment by ThingsChange** [comment]
i can''t  write any words;help me

**Comment by nariba** [comment]
I use CentOS 7.4 and zsh v5.0.2 .
@nomike mentioned also, latest CentOS and REHL does not support zsh 5.0.8 (zsh 5.0.2 is supported and latest version...).
So users can't use latest package without source code building.

**Comment by ericfreese** [comment]
Ok. Given how common these old versions are, I'm intending to start officially supporting zsh versions back to v4.3.11 (as zsh-syntax-highlighting does).

I've pushed up a branch [fixes/support_older_versions](https://github.com/zsh-users/zsh-autosuggestions/tree/fixes/support_older_versions) that attempts to run CI on these older versions, though I haven't got it working quite right yet...

PRs to that branch to improve support for older versions would be appreciated.

Thanks all

**Comment by Leetsong** [comment]
Do anyone know how to exit zsh? I cannot write any words (on my VPS server) now.

**Comment by Leetsong** [comment]
Have you fixed it? @ThingsChange

**Comment by SunMar** [comment]
If zsh is your default shell you can't exit it (or rather it will terminate your connection). If you use ssh you can add a command to connect to your server: `ssh [user@]host bash -l`, that should give you a shell in which you can revert the plugin to the previous version.

**Comment by nomike** [comment]
How about 
```
if $zshversion <= 5.0.8; then
    #old_code
else
    #new_code
fi
``` (pseudocode)

Or maybe that's already what branch fixes/support_older_versions is about.

**Comment by Leetsong** [comment]
Thanks @SunMar but when I used it, the server just complained`stdin: is not a tty`.

@nomike That is a workaround, but just for when I can access my server, but by now, I can't. :-(

**Comment by SunMar** [comment]
@Leetsong try it without the `-l`, it may looks like it is not doing anything but if you type something and hit enter you see you have a bash shell

**Comment by Leetsong** [comment]
@SunMar I've already tried before, but nothing happened except that it got stuck....

**Comment by SunMar** [comment]
@Leetsong are you sure it got stuck? try typing `ls -l`, for me it's just a bash shell but with an empty prompt so it looks like it's stuck but you can actually enter commands as normal

**Comment by Leetsong** [comment]
@SunMar Yes, I am sure. Other commands works fine. I am now trying to use `sed` to delete words that do something with zsh-autosuggestions. :-) Thanks for your inspiring.

**Comment by SunMar** [comment]
@Leetsong perhaps there is a way to start zsh without loading .zshrc or loading a non-existing file instead ? good luck :)

**Comment by ericfreese** [comment]
@SunMar @Leetsong You can run `zsh` with the `-f` flag to tell it not to source `.zshrc` at startup: `zsh -f`

**Comment by nomike** [comment]
@Leetsong:
You can try to allocate a pseudo terminal by executing
`ssh -t user@hostname 'bash'`
or if this fails reset the login shell using
`ssh user@hostname 'chsh -s /usr/bin/bash`

**Comment by Leetsong** [comment]
@nomike Thanks. You'v tought me another skill :-). And I have successfully fixed it using `sed`.

**Comment by Leetsong** [comment]
@ericfreese Thanks

**Comment by lnunno** [comment]
For antigen users I pinned to an older version to fix my issues e.g.:

`antigen bundles zsh-users/zsh-autosuggestions@v0.4.0`

Tested on zsh 5.0.2

**Comment by ericfreese** [comment]
Ok, this should be fixed on `master` and released as v0.4.2.
